{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "file = \"./consensus/aligned.fa\"\n",
    "records = list(SeqIO.parse(file, \"fasta\"))\n",
    "records = {r.id: r.seq for r in records}\n",
    "# make uppercase\n",
    "records = {k: v.upper() for k, v in records.items()}\n",
    "ref_file = \"ref.fa\"\n",
    "ref = list(SeqIO.parse(ref_file, \"fasta\"))[0]\n",
    "ref = ref.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "file = \"./sra_metadata.csv\"\n",
    "cols = [\"id\",\"biosample\",\"sample_name\",\"drug\"]\n",
    "# no header, usecols to select columns\n",
    "df = pd.read_csv(file, header=None, usecols=[0,1,2,3], names=cols)\n",
    "# sample name is study_name/detailed_sample_name/country, split these up\n",
    "df[\"study_name\"] = df[\"sample_name\"].str.split(\"/\").str[0]\n",
    "df[\"detailed_sample_name\"] = df[\"sample_name\"].str.split(\"/\").str[1]\n",
    "df[\"country\"] = df[\"sample_name\"].str.split(\"/\").str[2]\n",
    "\n",
    "# detailed sample name is individualid_day, split these up\n",
    "df[\"individualid\"] = df[\"detailed_sample_name\"].str.split(\"_\").str[0]\n",
    "df[\"day\"] = df[\"detailed_sample_name\"].str.split(\"_\").str[1]\n",
    "\n",
    "# drop detailed sample name\n",
    "df = df.drop(columns=[\"detailed_sample_name\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_mutations(start_seq, end_seq):\n",
    "    \"\"\"Find mutations between two sequences. If either sequence has \"n\" or \"-\" do not call this a mutation\"\"\"\n",
    "    mutations = []\n",
    "    to_ref = 0\n",
    "    not_to_ref = 0\n",
    "    for i in range(len(start_seq)):\n",
    "        if start_seq[i] != end_seq[i] and start_seq[i] != \"N\" and end_seq[i] != \"N\" and start_seq[i] != \"-\" and end_seq[i] != \"-\":\n",
    "            # get the 3nt centered on the mutation from the reference\n",
    "            context = ref[i-1:i+2]\n",
    "            mutations.append((i, start_seq[i], end_seq[i], context))\n",
    "            # check if is a mutation to reference\n",
    "            if end_seq[i] == ref[i]:\n",
    "                to_ref += 1\n",
    "            else:\n",
    "                not_to_ref += 1\n",
    "    if to_ref>0 and to_ref >= not_to_ref:\n",
    "        print(\"many mutations to reference, ignoring pair\")\n",
    "        return []\n",
    "    return mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "201 [] molnupiravir\n",
      "202 [] molnupiravir\n",
      "203 [] placebo\n",
      "204 [] placebo\n",
      "205 [] molnupiravir\n",
      "206 [] placebo\n",
      "207 [(13501, 'C', 'T', Seq('CCG'))] molnupiravir\n",
      "208 [] placebo\n",
      "209 [] placebo\n",
      "d5 not in records SRR19915106\n",
      "211 [(10424, 'G', 'A', Seq('GGT')), (16522, 'C', 'T', Seq('ACA')), (18815, 'G', 'A', Seq('TGT')), (23120, 'C', 'T', Seq('GCA')), (28115, 'G', 'A', Seq('CGA'))] molnupiravir\n",
      "212 [] placebo\n",
      "213 [] placebo\n",
      "214 [] molnupiravir\n",
      "215 [] molnupiravir\n",
      "216 [] placebo\n",
      "217 [] molnupiravir\n",
      "218 [] placebo\n",
      "219 [] placebo\n",
      "220 [(18651, 'C', 'T', Seq('GCG'))] molnupiravir\n",
      "221 [] molnupiravir\n",
      "222 [] placebo\n",
      "d5 not in records SRR19914624\n",
      "224 [] placebo\n",
      "225 [] placebo\n",
      "226 [(1230, 'G', 'T', Seq('AGT')), (1417, 'G', 'T', Seq('TGC')), (12356, 'T', 'C', Seq('ACA')), (14211, 'G', 'A', Seq('TGA')), (15445, 'G', 'A', Seq('TGT')), (20828, 'A', 'T', Seq('TAA')), (28094, 'T', 'A', Seq('TAA')), (28870, 'T', 'G', Seq('AGG'))] molnupiravir\n",
      "227 [] placebo\n",
      "228 [(750, 'C', 'T', Seq('ACA')), (1018, 'C', 'T', Seq('GCA')), (4087, 'G', 'A', Seq('TGA')), (17971, 'G', 'A', Seq('AGA')), (24067, 'C', 'T', Seq('ACA')), (25183, 'C', 'T', Seq('GCA'))] molnupiravir\n",
      "229 [] placebo\n",
      "230 [(15893, 'G', 'A', Seq('TGC')), (16238, 'T', 'C', Seq('CTG')), (17018, 'G', 'A', Seq('AGT')), (29349, 'A', 'G', Seq('CAT'))] molnupiravir\n",
      "231 [] placebo\n",
      "232 [] molnupiravir\n",
      "233 [] placebo\n",
      "234 [] molnupiravir\n",
      "235 [] placebo\n",
      "236 [] molnupiravir\n",
      "237 [(19567, 'T', 'C', Seq('GTT'))] molnupiravir\n",
      "238 [] placebo\n",
      "239 [(25537, 'G', 'T', Seq('GGC'))] placebo\n",
      "240 [] molnupiravir\n",
      "241 [] placebo\n",
      "242 [] molnupiravir\n",
      "243 [] molnupiravir\n",
      "244 [(27480, 'C', 'T', Seq('ACT'))] placebo\n",
      "245 [(11513, 'C', 'T', Seq('ACT')), (26872, 'C', 'T', Seq('ACA'))] molnupiravir\n",
      "246 [] placebo\n",
      "247 [] placebo\n",
      "248 [] molnupiravir\n",
      "249 [(5268, 'A', 'G', Seq('AAT')), (21639, 'T', 'C', Seq('CTG'))] molnupiravir\n",
      "250 [] placebo\n",
      "251 [] molnupiravir\n",
      "252 [(255, 'G', 'A', Seq('CGA')), (4064, 'C', 'T', Seq('TCT')), (6779, 'T', 'C', Seq('TTC')), (11019, 'C', 'T', Seq('TCA'))] molnupiravir\n",
      "253 [] placebo\n",
      "254 [] placebo\n",
      "255 [] molnupiravir\n",
      "256 [(10890, 'T', 'C', Seq('GTA')), (12126, 'T', 'C', Seq('ATG')), (19671, 'C', 'T', Seq('ACA')), (19682, 'A', 'G', Seq('AAG')), (20220, 'C', 'T', Seq('ACA'))] placebo\n",
      "many mutations to reference, ignoring pair\n",
      "257 [] placebo\n",
      "258 [] molnupiravir\n",
      "259 [] molnupiravir\n",
      "260 [] molnupiravir\n",
      "many mutations to reference, ignoring pair\n",
      "262 [] placebo\n",
      "263 [] placebo\n",
      "264 [] molnupiravir\n",
      "265 [] placebo\n",
      "266 [(7620, 'T', 'C', Seq('GTG')), (28813, 'C', 'T', Seq('TCA'))] molnupiravir\n",
      "267 [] placebo\n",
      "268 [] molnupiravir\n",
      "269 [(22984, 'G', 'A', Seq('GGC'))] placebo\n",
      "many mutations to reference, ignoring pair\n",
      "270 [] placebo\n",
      "many mutations to reference, ignoring pair\n",
      "271 [] molnupiravir\n",
      "272 [] molnupiravir\n",
      "273 [] molnupiravir\n",
      "274 [] placebo\n",
      "275 [] molnupiravir\n",
      "276 [] placebo\n",
      "277 [] molnupiravir\n",
      "278 [] molnupiravir\n",
      "279 [] placebo\n",
      "280 [] placebo\n",
      "281 [] placebo\n",
      "282 [] placebo\n",
      "283 [] molnupiravir\n",
      "284 [] molnupiravir\n",
      "many mutations to reference, ignoring pair\n",
      "285 [] placebo\n",
      "286 [] molnupiravir\n",
      "287 [] placebo\n",
      "288 [] placebo\n",
      "290 [(6949, 'A', 'G', Seq('TAT'))] molnupiravir\n",
      "291 [] placebo\n",
      "many mutations to reference, ignoring pair\n",
      "292 [] molnupiravir\n",
      "many mutations to reference, ignoring pair\n",
      "293 [] placebo\n",
      "294 [(4867, 'G', 'T', Seq('TGT'))] molnupiravir\n",
      "295 [(3970, 'G', 'T', Seq('TGT')), (8507, 'G', 'T', Seq('AGA')), (28041, 'G', 'T', Seq('GGA'))] placebo\n",
      "296 [] molnupiravir\n",
      "297 [] molnupiravir\n",
      "298 [] placebo\n",
      "d5 not in records SRR19914710\n",
      "300 [] molnupiravir\n",
      "301 [] placebo\n",
      "302 [] molnupiravir\n",
      "303 [] placebo\n",
      "304 [] placebo\n",
      "many mutations to reference, ignoring pair\n",
      "305 [] molnupiravir\n",
      "306 [(11376, 'T', 'C', Seq('GTG')), (22217, 'G', 'T', Seq('GGT'))] molnupiravir\n",
      "307 [] molnupiravir\n",
      "308 [] placebo\n",
      "d1 not in records SRR19915055\n",
      "310 [] molnupiravir\n",
      "311 [] placebo\n",
      "312 [] molnupiravir\n",
      "313 [] placebo\n",
      "314 [] molnupiravir\n",
      "315 [] molnupiravir\n",
      "316 [] placebo\n",
      "317 [] placebo\n",
      "318 [] placebo\n",
      "319 [] molnupiravir\n",
      "320 [(5125, 'T', 'C', Seq('TTT')), (17601, 'G', 'T', Seq('GGT')), (17667, 'G', 'T', Seq('GGG')), (18123, 'G', 'T', Seq('TGT')), (19281, 'G', 'T', Seq('TGA')), (24139, 'G', 'T', Seq('TGT'))] molnupiravir\n",
      "321 [] placebo\n",
      "322 [] placebo\n",
      "323 [(240, 'C', 'T', Seq('TCG'))] molnupiravir\n",
      "324 [] molnupiravir\n",
      "325 [(4066, 'G', 'A', Seq('TGC')), (21341, 'T', 'C', Seq('ACA')), (25176, 'G', 'A', Seq('AGT')), (26347, 'C', 'T', Seq('ACT'))] molnupiravir\n",
      "326 [] placebo\n",
      "327 [(23598, 'T', 'G', Seq('ATT'))] molnupiravir\n",
      "many mutations to reference, ignoring pair\n",
      "328 [] placebo\n",
      "329 [] placebo\n",
      "330 [] molnupiravir\n",
      "331 [] placebo\n",
      "332 [] molnupiravir\n",
      "333 [] molnupiravir\n",
      "334 [] placebo\n",
      "many mutations to reference, ignoring pair\n",
      "335 [] placebo\n",
      "336 [] molnupiravir\n",
      "337 [] placebo\n",
      "338 [(23598, 'T', 'G', Seq('ATT'))] molnupiravir\n",
      "339 [] placebo\n",
      "340 [(13317, 'A', 'G', Seq('AAA'))] molnupiravir\n",
      "many mutations to reference, ignoring pair\n",
      "341 [] molnupiravir\n",
      "342 [] placebo\n",
      "343 [(1945, 'G', 'A', Seq('TGT')), (2579, 'A', 'G', Seq('GAA')), (8325, 'C', 'T', Seq('ACC')), (13182, 'G', 'T', Seq('AGG')), (14324, 'C', 'T', Seq('ACC')), (14830, 'G', 'A', Seq('TGT')), (18311, 'C', 'T', Seq('TCG')), (21661, 'C', 'T', Seq('ACG')), (23145, 'A', 'G', Seq('AAA')), (28319, 'C', 'T', Seq('ACG'))] molnupiravir\n",
      "344 [] placebo\n",
      "345 [] placebo\n",
      "346 [] molnupiravir\n",
      "347 [] placebo\n",
      "many mutations to reference, ignoring pair\n",
      "348 [] molnupiravir\n",
      "many mutations to reference, ignoring pair\n",
      "349 [] placebo\n",
      "350 [(10446, 'G', 'A', Seq('GGC'))] molnupiravir\n",
      "351 [(21769, 'G', 'A', Seq('TGT'))] molnupiravir\n",
      "352 [] placebo\n",
      "353 [] placebo\n",
      "354 [(12669, 'T', 'C', Seq('CTG'))] molnupiravir\n",
      "355 [] molnupiravir\n",
      "356 [] placebo\n",
      "357 [(14132, 'T', 'C', Seq('TTC')), (14194, 'C', 'T', Seq('ACT'))] molnupiravir\n",
      "d1 not in records SRR19914994\n",
      "359 [(5885, 'C', 'T', Seq('ACA')), (21499, 'T', 'C', Seq('ATA'))] placebo\n",
      "360 [(15678, 'G', 'T', Seq('CGC'))] molnupiravir\n",
      "361 [] placebo\n",
      "many mutations to reference, ignoring pair\n",
      "362 [] molnupiravir\n",
      "363 [] placebo\n",
      "364 [] molnupiravir\n",
      "365 [] molnupiravir\n",
      "366 [(3691, 'G', 'T', Seq('AGT')), (14136, 'G', 'A', Seq('TGT')), (21775, 'G', 'A', Seq('TGG')), (24555, 'A', 'G', Seq('CAG')), (25142, 'A', 'G', Seq('AAT')), (25476, 'G', 'A', Seq('TGT')), (29148, 'C', 'T', Seq('TCA'))] molnupiravir\n",
      "many mutations to reference, ignoring pair\n",
      "367 [] placebo\n",
      "368 [(240, 'C', 'T', Seq('TCG')), (29121, 'A', 'G', Seq('AAG'))] placebo\n",
      "369 [] molnupiravir\n",
      "370 [] placebo\n",
      "371 [] molnupiravir\n",
      "372 [] placebo\n",
      "373 [(240, 'C', 'T', Seq('TCG'))] placebo\n",
      "374 [(2158, 'G', 'T', Seq('TGA')), (11561, 'G', 'A', Seq('TGC')), (21692, 'G', 'T', Seq('AGA'))] molnupiravir\n",
      "375 [(16723, 'T', 'C', Seq('TTA'))] placebo\n",
      "376 [] molnupiravir\n",
      "377 [(7621, 'G', 'A', Seq('TGT'))] molnupiravir\n",
      "378 [] placebo\n",
      "d5 not in records SRR19914803\n",
      "380 [(240, 'C', 'T', Seq('TCG'))] placebo\n",
      "[(240, 4), (23598, 2), (13501, 1), (10424, 1), (16522, 1), (18815, 1), (23120, 1), (28115, 1), (18651, 1), (1230, 1)]\n"
     ]
    }
   ],
   "source": [
    "def transition_or_tranvsversion(start,end):\n",
    "    \"\"\"Return transition or transversion\"\"\"\n",
    "    if start == \"A\" and end == \"G\":\n",
    "        return \"transition\"\n",
    "    if start == \"G\" and end == \"A\":\n",
    "        return \"transition\"\n",
    "    if start == \"C\" and end == \"T\":\n",
    "        return \"transition\"\n",
    "    if start == \"T\" and end == \"C\":\n",
    "        return \"transition\"\n",
    "    return \"transversion\"\n",
    "\n",
    "by_individual = df.groupby(\"individualid\")\n",
    "from collections import Counter, defaultdict\n",
    "site_counts = Counter()\n",
    "muts_by_drug = defaultdict(list)\n",
    "\n",
    "drug_counts = Counter()\n",
    "\n",
    "for individual, group in by_individual:\n",
    "    # try to find a d1 seq\n",
    "    d1 = group[group[\"day\"] == \"d1\"]\n",
    "    if len(d1) == 0:\n",
    "        continue\n",
    "    # try to find a d5 seq\n",
    "    d5 = group[group[\"day\"] == \"d5\"]\n",
    "    if len(d5) == 0:\n",
    "        continue\n",
    "\n",
    "    d3 = group[group[\"day\"] == \"d3\"]\n",
    "    # check the sequences are in records, otherwise print error and skip\n",
    "    if d1[\"id\"].values[0] not in records:\n",
    "\n",
    "        print(\"d1 not in records\", d1[\"id\"].values[0])\n",
    "        if d3[\"id\"].values[0] in records:\n",
    "            d1 = d3\n",
    "            print(\"using d3 instead\")\n",
    "        else:\n",
    "            continue\n",
    "    if d5[\"id\"].values[0] not in records:\n",
    "        print(\"d5 not in records\", d5[\"id\"].values[0])\n",
    "        continue\n",
    "    drug = d1[\"drug\"].values[0]\n",
    "    drug_counts[drug] += 1\n",
    "    # find mutations\n",
    "    mutations = find_mutations(records[d1[\"id\"].values[0]], records[d5[\"id\"].values[0]])\n",
    "    # print mutations\n",
    "    print(individual, mutations, drug)\n",
    "    for m in mutations:\n",
    "        site_counts[m[0]] += 1\n",
    "        muts_by_drug[drug].append(m)\n",
    "\n",
    "# print the top 10 sites\n",
    "print(site_counts.most_common(10))\n",
    "# count the number of each (start,end) pair for each drug, and print in reverse order\n",
    "\n",
    "#find sites with more than one count\n",
    "multiple_times = {k for k,v in site_counts.items() if v > 1}\n",
    "# filter these away from the muts_by_drug\n",
    "for drug, muts in muts_by_drug.items():\n",
    "    muts_by_drug[drug] = [m for m in muts if m[0] not in multiple_times]\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\\TOTALS ANALYSIS\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/29/cygttqzs4gq3vzhzg230d9pr0000gn/T/ipykernel_15487/706917553.py:9: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "/var/folders/29/cygttqzs4gq3vzhzg230d9pr0000gn/T/ipykernel_15487/706917553.py:9: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "/var/folders/29/cygttqzs4gq3vzhzg230d9pr0000gn/T/ipykernel_15487/706917553.py:9: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "/var/folders/29/cygttqzs4gq3vzhzg230d9pr0000gn/T/ipykernel_15487/706917553.py:9: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "/var/folders/29/cygttqzs4gq3vzhzg230d9pr0000gn/T/ipykernel_15487/706917553.py:9: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "/var/folders/29/cygttqzs4gq3vzhzg230d9pr0000gn/T/ipykernel_15487/706917553.py:9: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "/var/folders/29/cygttqzs4gq3vzhzg230d9pr0000gn/T/ipykernel_15487/706917553.py:9: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "/var/folders/29/cygttqzs4gq3vzhzg230d9pr0000gn/T/ipykernel_15487/706917553.py:9: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "/var/folders/29/cygttqzs4gq3vzhzg230d9pr0000gn/T/ipykernel_15487/706917553.py:9: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "/var/folders/29/cygttqzs4gq3vzhzg230d9pr0000gn/T/ipykernel_15487/706917553.py:9: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "/var/folders/29/cygttqzs4gq3vzhzg230d9pr0000gn/T/ipykernel_15487/706917553.py:9: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "/var/folders/29/cygttqzs4gq3vzhzg230d9pr0000gn/T/ipykernel_15487/706917553.py:9: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "/var/folders/29/cygttqzs4gq3vzhzg230d9pr0000gn/T/ipykernel_15487/706917553.py:9: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "/Users/theosanderson/miniconda3/envs/py310/lib/python3.10/site-packages/plotnine/ggplot.py:718: PlotnineWarning: Saving 5 x 5 in image.\n",
      "/Users/theosanderson/miniconda3/envs/py310/lib/python3.10/site-packages/plotnine/ggplot.py:719: PlotnineWarning: Filename: mutations_by_drug.pdf\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# divide the G->A mutations by their context for each drug\n",
    "print(\"\\n\\TOTALS ANALYSIS\")\n",
    "by_mut_type = {k: Counter([f\"{m[1]}>{m[2]}\" for m in muts]).most_common() for k, muts in muts_by_drug.items()}\n",
    "from plotnine import *\n",
    "import pandas as pd\n",
    "df = pd.DataFrame()\n",
    "for drug, muts in by_mut_type.items():\n",
    "    for mut in muts:\n",
    "        df = df.append({\"drug\": drug\n",
    "                        , \"mutation\": mut[0]\n",
    "                        , \"count\": mut[1]\n",
    "                          }, ignore_index=True)\n",
    "# add in the zero values with expand_grid\n",
    "from itertools import product\n",
    "df = df.merge(pd.DataFrame(list(product(df[\"drug\"].unique(), df[\"mutation\"].unique())), columns=[\"drug\", \"mutation\"]), how=\"outer\")\n",
    "df[\"count\"] = df[\"count\"].fillna(0)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# plot the mutations by drug with bar\n",
    "p = (ggplot(df, aes(fill=\"drug\", y=\"count\", x=\"mutation\"))\n",
    "    + geom_bar(stat=\"identity\", position=\"dodge\")\n",
    "    + theme_classic() + labs(x=\"Mutation type\", y=\"Times observed\", fill = \"Drug\"))\n",
    "\n",
    "p.save(\"mutations_by_drug.pdf\", width=5, height=5, units=\"in\", dpi=300)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def is_mut_of_interest(start,end, context):\n",
    "    # g>a in the context of tgt\n",
    "    if start == \"G\" and end == \"A\" and \"T\" in context:\n",
    "        return True\n",
    "    # c>t in the context of aca\n",
    "    if start == \"C\" and end == \"T\" and \"A\" in context:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "\n",
    "for drug, muts in muts_by_drug.items():\n",
    "    print(drug)\n",
    "    print(Counter([m[1:3] for m in muts]).most_common())\n",
    "    print(Counter([m[3] for m in muts]).most_common())\n",
    "    print(Counter([is_mut_of_interest(m[1],m[2],m[3]) for m in muts]).most_common())\n",
    "\n",
    "\n",
    "# print the number of samples for each drug\n",
    "print(drug_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 29903/29903 [00:00<00:00, 151984.44it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGG 0.3890577507598784 0.6109422492401215\n",
      "GGT 0.16079295154185022 0.8392070484581498\n",
      "CGA 0.3157894736842105 0.6842105263157895\n",
      "TGT 0.2867132867132867 0.7132867132867133\n",
      "AGA 0.21818181818181817 0.7818181818181819\n",
      "TGG 0.17870036101083034 0.8212996389891697\n",
      "GGC 0.17937219730941703 0.820627802690583\n",
      "CGG 0.2894736842105263 0.7105263157894737\n",
      "TGC 0.1809872029250457 0.8190127970749543\n",
      "AGT 0.35700197238658776 0.6429980276134122\n",
      "CGC 0.29896907216494845 0.7010309278350515\n",
      "CGT 0.2807017543859649 0.7192982456140351\n",
      "TGA 0.18571428571428572 0.8142857142857143\n",
      "GGA 0.20921985815602837 0.7907801418439716\n",
      "AGC 0.23920265780730898 0.760797342192691\n",
      "GGG 0.373134328358209 0.6268656716417911\n",
      "5863 total G mutations\n",
      "5863 total Gs calculated another way\n",
      "AGG 0.08982456140350877 201\n",
      "GGT 0.051228070175438595 381\n",
      "CGA 0.021052631578947368 65\n",
      "TGT 0.1726315789473684 612\n",
      "AGA 0.09263157894736843 473\n",
      "TGG 0.06947368421052631 455\n",
      "GGC 0.028070175438596492 183\n",
      "CGG 0.015438596491228071 54\n",
      "TGC 0.06947368421052631 448\n",
      "AGT 0.1270175438596491 326\n",
      "CGC 0.020350877192982456 68\n",
      "CGT 0.03368421052631579 123\n",
      "TGA 0.08210526315789474 513\n",
      "GGA 0.041403508771929824 223\n",
      "AGC 0.05052631578947368 229\n",
      "GGG 0.03508771929824561 84\n"
     ]
    }
   ],
   "source": [
    "def translate_codon(codon):\n",
    "    lookup = dict(zip([\"TTT\", \"TTC\", \"TTA\", \"TTG\", \"CTT\", \"CTC\", \"CTA\", \"CTG\", \"ATT\", \"ATC\", \"ATA\", \"ATG\", \"GTT\", \"GTC\", \"GTA\", \"GTG\", \"TCT\", \"TCC\", \"TCA\", \"TCG\", \"CCT\", \"CCC\", \"CCA\", \"CCG\", \"ACT\", \"ACC\", \"ACA\", \"ACG\", \"GCT\", \"GCC\", \"GCA\", \"GCG\", \"TAT\", \"TAC\", \"TAA\", \"TAG\", \"CAT\", \"CAC\", \"CAA\", \"CAG\", \"AAT\", \"AAC\", \"AAA\", \"AAG\", \"GAT\", \"GAC\", \"GAA\", \"GAG\", \"TGT\", \"TGC\", \"TGA\", \"TGG\", \"CGT\", \"CGC\", \"CGA\", \"CGG\", \"AGT\", \"AGC\", \"AGA\", \"AGG\", \"GGT\", \"GGC\", \"GGA\", \"GGG\"], \"FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG\"))\n",
    "    return lookup[codon]\n",
    "from collections import Counter, defaultdict\n",
    "import tqdm\n",
    "file = \"hu1.gb\"\n",
    "from Bio import SeqIO\n",
    "records = list(SeqIO.parse(file, \"genbank\"))\n",
    "main = records[0]\n",
    "# find CDS features\n",
    "cds = [f for f in main.features if f.type == \"CDS\"]\n",
    "#print(cds)\n",
    "# for each position in the sequence, find the feature it is in\n",
    "\n",
    "context_to_type = defaultdict(Counter)\n",
    "\n",
    "for i in tqdm.tqdm(range(len(main.seq))):\n",
    "    my_val = main.seq[i]\n",
    "    one_based = i+1\n",
    "    this_cds = None\n",
    "    for f in cds:\n",
    "        if i >= f.location.start.position and i <= f.location.end.position:\n",
    "            this_cds = f\n",
    "            break\n",
    "    if this_cds:\n",
    "        offset = i - this_cds.location.start.position\n",
    "        codon = offset // 3\n",
    "        codon_string = main.seq[this_cds.location.start.position + codon*3:this_cds.location.start.position + codon*3+3]\n",
    "        my_position = offset % 3\n",
    "    else:\n",
    "        codon = None\n",
    "        codon_string = None\n",
    "        my_position = None\n",
    "    cds_name = this_cds.qualifiers[\"gene\"][0] if this_cds else None\n",
    "    context = main.seq[i-1:i+2]\n",
    "    if my_val == \"G\":\n",
    "        if codon is None:\n",
    "            type = \"syn\"\n",
    "        else:\n",
    "            mutated_string= list(codon_string)\n",
    "            mutated_string[my_position] = \"A\"\n",
    "            mutated_string = \"\".join(mutated_string)\n",
    "            if translate_codon(mutated_string) == translate_codon(codon_string):\n",
    "                type = \"syn\"\n",
    "            else:\n",
    "                type = \"nonsyn\"\n",
    "        context_to_type[context][type] += 1\n",
    "    \n",
    "    #print(i, cds_name, my_val, one_based, codon, codon_string, my_position)\n",
    "\n",
    "# print proportions for each context\n",
    "\n",
    "grand_total = 0\n",
    "for context, counts in context_to_type.items():\n",
    "    total = sum(counts.values())\n",
    "    grand_total += total\n",
    "    print(context, counts[\"syn\"]/total, counts[\"nonsyn\"]/total)\n",
    "\n",
    "print(grand_total, \"total G mutations\")\n",
    "print(len([i for i in main.seq if i == \"G\"]), \"total Gs calculated another way\")\n",
    "\n",
    "total_syn = sum([counts[\"syn\"] for counts in context_to_type.values()])\n",
    "for context, counts in context_to_type.items():\n",
    "\n",
    "    print(context, counts[\"syn\"]/total_syn , counts[\"nonsyn\"])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py310",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8 (main, Nov 24 2022, 08:08:27) [Clang 14.0.6 ]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "c8d440e3959747bf1e99f498e2fad430a3012487f32fa421a32f347fc0778ddc"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
